### ISSUE TITLE
Quantum Computer does not work with ME Level Emitter

### ISSUE THREAD

**Issue by maxnoller** [pedroksl/AdvancedAE#5]
I tried setting up a level emitter to emit a redstone signal to craft Heavy Water from Mekanism in the ATM10 Modpack. It does not work when selecting the Quantum Computer as crafting CPU but works when using normal ae2 crafting cpus.

If you need any additional information or I can help in some way let me know!

**Comment by rjbowen97** [comment]
I am encountering this issue as well! I am playing as a part of All the Mods 10. Tried switching between "emit to craft" and "emit while crafting", but neither one works. Switching between these two settings does briefly turn on the level emitter though, but just for a split second

**Comment by pedroksl** [comment]
Hey, sorry for the long time to try and look at this issue.

I was trying to use the level emitter with the autocrafting setup that is in the ae2guide. It did bug out and never completed the craft. However that also happened with a standard ae2 computer. I also don't see how the quantum computer would be any different from a standard cpu in this case (except that it is consuming the entire cpu, which it shouldn't do, ever). I'll look into fixing that part of the problem and then take a look at why it is hanging.

**Comment by sp1rr1t** [comment]
Hello everyone, I need help. My quantum computer does not work at all, it becomes a full-fledged structure and is visible on the network, but as soon as a request request appears, it just freezes and does nothing, while I cannot cancel the crafting. I play on ATM10

**Comment by pedroksl** [comment]
If you're using Level Emitters, unfortunately, that bug still hasn't been solved. I've stepped through the code multiple times and everything looks to be working perfectly, and then nothing happens. The only thing I can suggest, for now, is to remove level emitter and use other methods (like requesters or the quantum crafter) and break the quantum computer and place it again, even though this may not fix it due to nbt data. Unfortunately, this is all I have for now...

If there are no level emitters in the network, there might be more troubleshooting possible. But I would suggest creating a new issue instead of reusing this one.

**Comment by ehrenfriedtim** [comment]
Maybe you could provide us an „Advance Level Emitter“, with the same Featureset, but tested/working with the quantum CPU ?

**Comment by Kribylet** [comment]
Having this issue too using an ME Level Emitter with craft setting, observed this:

Changing the emit style caused the emitter to trigger briefly.
Replacing an oversized ME interface on the same small ME cable started the emitter for a longer period of time, but not the duration of the craft.

**Comment by DenysChuhlib** [comment]
I also play ATM10 and when using a quantum computer, only after updating the network does the emitter turn on (operation mode during crafting) for a couple of seconds and then turn off, but when using a regular processor (from AE2) it works normally.

**Comment by Mithi83** [comment]
I tried to debug this from the AE2 side of things in https://github.com/AppliedEnergistics/Applied-Energistics-2/issues/8251

My current finding is that somehow `lastModifiedOnTick` in `AdvCraftingCPULogic` is always 0 when using a Quantum Computer Core. I haven't had a deeper look at your code, but you seem to have copied a lot from AE2 and added some sort of wrapper for the individual crafting jobs you spawn dynamically. As far as I can see the intention of the AE2 code is that whenever a craft finishes and the result is returned to the appropriate crafting CPU this variable is updated. I don't know if you pass that event on to the dynamically spawned `AdvCraftingCPULogic` or if you handle this centrally for all crafting jobs running on a Quantum Computer Core. Maybe this can help a bit with your debugging efforts.

**Comment by pedroksl** [comment]
Thank you for taking your time trying to debug this issue. As you can see, this is a long standing bug, with no solution, still. You probably found that out by now, but most of this is handled in a mixin I have in the CraftingService. `onServerEndTick` is a very difficult method to inject because there are multiple places where changes are required and I would never want to simply override it to implement what I needed. What I did instead was inject at the end and try to replicate the required parts for the quantum computer to work properly. I still believe the issue might be somewhere in there, but I fail to see where. I'll try to do some more testing when I get some more free time. Again, thanks for taking your time, you definitely did not have to, I appreaciate it.

**Comment by Mithi83** [comment]
Thanks for the mixin hint, I wouldn't have found that one on my own any time soon. I'm not familiar with this construct, but what I gather from the wording there I can make an educated guess at what it does. This approach seems a bit fragile to me, but unfortunately I have no good idea how to improve it easily. You basically have a copy of the code that will run after AE2 handles all normal Crafting CPUs. This might lead to interesting effects (i.e. unexpected behavior) when crafting with normal AE2 CPUs and Quantum CPUs at the same time as events are sent on partial information at different stages.

Ideally you would hook into one singular part of that function (constructing the `currentlyCrafting` set) rather than doing it at the end and having to duplicate the notification part in the process. I'm not sure if mixins can target that precisely but if not you'd benefit from a refactor of the AE2 implementation that puts the code you want to amend in a separate function that you can then hook into with a more precise mixin. Looking at your code you probably only need the `for (var cluster : this.advancedAE$advCraftingCPUClusters)` if you could run your mixin before executing the notification stuff.

Comparing the copy to the original code revealed an interesting difference though and a quick test showed that it improved the situation a bit. You have `var previouslyCrafting = this.currentlyCrafting;` in there and that will probably not do what you want it to do, as you effectively have the same HashSet that you access via `previouslyCrafting` and `currentlyCrafting`. Creating the set difference below will thus always result in an empty set and thus never trigger the `onRequestChange`. Correct me if I'm wrong, Java is not my primary programming language.

I replaced it like this with code found in AE2, which does a deep copy instead:
```java
@Inject(method = "onServerEndTick", at = @At("TAIL"))
private void tickAdvClusters(CallbackInfo ci) {
    Set<AEKey> previouslyCrafting = currentlyCrafting.isEmpty() ? Set.of() : new HashSet<>(currentlyCrafting);
```

I now get two breakpoint triggers, one when the first craft starts coming from your mixin and one after the last craft stopped, coming from the original AE2 CraftingService.onServerEndTick. In contrast to a normal AE2 Crafting CPU, the ME Level Emitter stays on continuously throughout all the crafts, whereas in the AE2 case it flickers rapidly between on and off for each individual craft. I have not yet understood why this happens, but maybe we're one step closer.

**Comment by pedroksl** [comment]
Interesting find. My main language is also not Java, I come from C++. It's sometimes hard to wrap my head around everything being a pointer.

The main thing I wanted out of that mixin was to tick my crafting the jobs. All the rest is just me trying to mimic what normal jobs do when they tick. The easiest approach, by far, would be if AE2 had a separate method that did the ticking. I could mixin that method and simply add my own and change the returned "latestChange" appropriately, and the integration would be seemingless, as far as I know.

I tried so much to mixin right after that for loop but couldn't find the appropriate injection method to do so. This ended up in me landing in the current approach of trying to mimic everything. That said, maybe a PR to change that implementation and create a separate method is the easiest solution. But I would like some form of confirmation it works before doing so. I could maybe try to compile a different AE2 version with that change and mixin into that.

**Comment by Mithi83** [comment]
Nice, I'm mainly using C++ as well, at least professionally.

Mystery solved I guess: Your mixin only ever adds to the `currentlyCrafting` set, but never resets it. That is only done in the AE2 code, but that has been refactored here (https://github.com/AppliedEnergistics/Applied-Energistics-2/commit/1a277e6db0e11720a33298e7c6da7aa2b8db3dc9#diff-60e456341fe32f3a5a00a12e8a6d1140db49be5fc7a9e43c064c59bea4c4406eR163) to only happen if a change was registered while ticking. As your dynamic jobs don't tick in the main loop but your mixin, they can't trigger the reset in the AE2 code. Also you likely copied that code before that refactor was made, so I suggest looking at that refactor and try to figure out how that impacts your implementation. It is possible that your mixin worked before that refactoring.

A quickfix for some cases (i.e. you only ever craft with Quantum CPUs and not with normal AE2 CPUs) would be to change it like this:

```java
    @Inject(method = "onServerEndTick", at = @At("TAIL"))
    private void tickAdvClusters(CallbackInfo ci) {
        Set<AEKey> previouslyCrafting = currentlyCrafting.isEmpty() ? Set.of() : new HashSet<>(currentlyCrafting);
        this.currentlyCrafting.clear();
        for (var cluster : this.advancedAE$advCraftingCPUClusters) {
``` 

This will probably break a mixed setup in some way, which is why I can't recommend it as a proper fix or PR. My test setup with Quantum CPU only now has working ME Level Emitters for each crafting step.

You could try to use a different HashSet for your crafting CPUs and not add it to `this.currentlyCrafting`, e.g. introduce `this.currentlyCraftingAdvAE` and use that with a proper reset each tick. However, you still need to add it to `this.currentlyCrafting` because other methods like `CraftingService.isRequesting` or `CraftingService.isRequestingAny` check that variable as well. Or you need to add mixins for that as well.

**Comment by iczero** [comment]
I think it should be possible to inject into `onServerEndTick` more precisely, which could avoid some hacks. `currentlyCrafting.clear()` is called in an if block which only happens if the previous code decides that any CPU has updated, so more precise inject might fix that. I will try that.

**Comment by Mithi83** [comment]
Nice one! I tried your PR in my test setup and it seems to work for ME Level Emitters with Crafting Cards and a Quantum Computer Core. I haven't taken a look at mixed setups with both regular and quantum CPUs yet, but I'm quite confident, that it *should* work. 

Thanks for helping us out here! Your approach is basically what I roughly had in mind but I wasn't able to write the code for that. If you wouldn't mind: You seem quite knowledgeable about this mixin stuff. Do you have a good reference/documentation for that, so I can learn a bit about it?

**Comment by iczero** [comment]
To be entirely honest, I'm not too familiar with mixins. In any case, here's a list of resources I've looked at:

Mixin wiki
- https://github.com/SpongePowered/Mixin/wiki/Advanced-Mixin-Usage---Callback-Injectors
- https://github.com/SpongePowered/Mixin/wiki/Advanced-Mixin-Usage---Redirect-Injectors
- https://github.com/SpongePowered/Mixin/wiki/Injection-Point-Reference

asm reference: https://asm.ow2.io/asm4-guide.pdf

Fabric wiki: 
- https://wiki.fabricmc.net/tutorial:mixin_examples
- https://wiki.fabricmc.net/tutorial:mixin_tips

I've also found it useful to read mixins from other mods.

**Comment by pedroksl** [comment]
The idea of using the getters for the first variables in the `if` code blocks right after the cpu iterators actually solved my initial issue of not finding the best place to inject into that method. Thank you for your contribution and also Mithi83 for spending the time and energy to help us debug this one.

**Comment by iczero** [comment]
Thank you as well for writing such a nice mod!

**Comment by Mithi83** [comment]
You‘re welcome. I‘d say it was a nice team effort and I even learned something. Thanks for the documentation links!
